Feat: Authenticate your Kong Gateway Amazon RDS database with AWS IAM (#1810)
Kong/developer.konghq.com: app/gateway/amazon-rds-authentication-with-aws-iam.md

---
title: "{{site.base_gateway}} Amazon RDS database authentication with AWS IAM"
content_type: reference
layout: reference

breadcrumbs:
  - /gateway/
products:
    - gateway

tags:
  - aws

min_version:
    gateway: '3.3'

description: "Learn how to use AWS Identity and Access Management (IAM) authentication to connect to the Amazon RDS database that you use for {{site.base_gateway}}"

related_resources:
  - text: "Install {{site.base_gateway}}"
    url: /gateway/install/

works_on:
  - on-prem
---

You can use AWS Identity and Access Management (IAM) authentication to connect to the Amazon RDS database that you use for {{site.base_gateway}}. This page describes how to use this feature to secure your database configurations and database connections.

> When you enable this feature, you don't need to use a password when you connect to a database instance. Instead, you use a temporary authentication token. Because AWS IAM manages the authentication externally, the database doesn't store user credentials. If you use Amazon RDS for {{site.base_gateway}}'s database, you can enable this feature on your running cluster. This ensures that you don't have to store database user credentials on both the {{site.base_gateway}} (`pg_password`) and RDS database side. 

Reviewer: Just to avoid repeating "this feature".

   {:.info}
   > **Notes:** 
   > * The database user assigned to the `rds_iam` role can only use the IAM database authentication.
   > * Make sure to create the database and grant the correct permissions to the database user you just created.

### Configuring AWS IAM authentication in {{site.base_gateway}}

{% navtabs "Configuration" %}
{% navtab "Environment variables" %}

To enable AWS IAM authentication in read-write and read-only mode, set the `KONG_PG_IAM_AUTH` environment variable to `on`: 

```bash
KONG_PG_IAM_AUTH=on
```

To enable AWS IAM authentication in read-only mode, you can set the following:

```bash
KONG_PG_IAM_AUTH=off # This variable can be omitted because off is the default value
KONG_PG_RO_IAM_AUTH=on
```

{% new_in 3.8 %} If you want to [assume a role](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html), also set the following environment variables:

```bash
# For read-write connections
KONG_PG_IAM_AUTH_ASSUME_ROLE_ARN=$ROLE_ARN
KONG_PG_IAM_AUTH_ROLE_SESSION_NAME=$ROLE_SESSION_NAME

> # Optional, specify the custom STS endpoint URL used for the IAM assume role

Reviewer: ```suggestion
# Optional: Specify a custom STS endpoint URL for IAM role assumption
```

# This value will override the default STS endpoint URL, which should be
# `https://sts.amazonaws.com`, or `https://sts.$REGION.amazonaws.com` if
# `AWS_STS_REGIONAL_ENDPOINTS` is set to `regional`(by default).

> # If you are not using a private VPC endpoint for STS service, you should
> # not specify this value

Reviewer: ```suggestion
# Only set this if you're using a private VPC endpoint for the STS service
```

KONG_PG_IAM_AUTH_STS_ENDPOINT_URL=$STS_ENDPOINT

# For read-only connections, if you need a different role than for read-write
KONG_PG_RO_IAM_AUTH_ASSUME_ROLE_ARN=$ROLE_ARN
KONG_PG_RO_IAM_AUTH_ROLE_SESSION_NAME=$ROLE_SESSION_NAME
# Optional, same as KONG_PG_IAM_AUTH_STS_ENDPOINT_URL
KONG_PG_RO_IAM_AUTH_STS_ENDPOINT_URL=$STS_ENDPOINT
```
{% endnavtab %}

{% navtab "Configuration file" %}
To enable AWS IAM authentication in read-write mode, set `pg_iam_auth` to `on` in [`kong.conf`](/gateway/configuration/):
```text
pg_iam_auth=on
```

To enable AWS IAM authentication in read-only mode, set `pg_ro_iam_auth` to `on`:
```text
pg_ro_iam_auth=on
```

{% new_in 3.8 %} If you want to [assume a role](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html), also set the following configuration parameters:

```bash
# For read-write connections
pg_iam_auth_assume_role_arn=$ROLE_ARN
pg_iam_auth_role_session_name=$ROLE_SESSION_NAME

> # Optional, specify the custom STS endpoint URL used for the IAM assume role

Reviewer: ```suggestion
# Optional: Specify a custom STS endpoint URL for IAM role assumption 
```

# This value will override the default STS endpoint URL, which should be
# `https://sts.amazonaws.com`, or `https://sts.$REGION.amazonaws.com` if
# `AWS_STS_REGIONAL_ENDPOINTS` is set to `regional`(by default).

> # If you are not using a private VPC endpoint for STS service, you should
> # not specify this value

Reviewer: ```suggestion
# Only set this if you're using a private VPC endpoint for the STS service 
```

pg_iam_auth_sts_endpoint_url=$STS_ENDPOINT

# For read-only connections, if you need a different role than for read-write
pg_ro_iam_auth_assume_role_arn=$ROLE_ARN
pg_ro_iam_auth_role_session_name=$ROLE_SESSION_NAME

> # Optional, same as `pg_iam_auth_sts_endpoint_url`

Reviewer: ```suggestion
# Optional: same as `pg_iam_auth_sts_endpoint_url`
```
